### DISCUSSION TITLE
Is there any way to specify service information per process within application for platform / orchestrator apps?

### DISCUSSION THREAD

**Discussion by rkolesnev** [open-telemetry/opentelemetry-java-instrumentation#5953] [Q&A] [answered]
Hi, 

I am doing some prototyping / discovery of using OpenTelemetry with Kafka products and am running into a restriction of Resource information being set on application level vs being able to specify / change those attributes per scope / process.
Specifically instrumenting things like Connect platform that runs as single java process but is not really a single service - more of an orchestrator / scheduler for different connector worker processes / tasks. Since the worker processes are loaded within same application / jvm instance - specifying service information per Connect platform instance doesn't fit right.

For my use case it would make sense to be able to set service information per connector - i.e. things like service name, version, namespace etc. I am instrumenting it with java-instrumentation agent using auto instrumentation approach and adding additional custom behaviour using extensions support.
Could anyone please share best practice for setting those attributes dynamically - or point me towards the discussion / general approach for doing platform rather than single service instrumentation - and gotchas / lessons learned - i would really appreciate that.

Of course i can use custom span attributes but problem then is with data export / visualization - for example using Jaeger - service name being that of Connect platform rather than individual connector services running on it.

Cheers,
Roman

**Comment by trask** [comment, marked as answer]
hi @rkolesnev! I think this is a common-ish use case, e.g. when using the Javaagent in an application server that hosts multiple applications.

we don't really have a pre-baked answer for this, but I hope some day we will

for now, I _think_ you could use a SpanProcessor to propagate custom "tenant" attributes down from parent to child, e.g. https://github.com/microsoft/ApplicationInsights-Java/blob/main/agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/InheritedAttributesSpanProcessor.java

and then write a SpanExporter wrapper that wraps the SpanData and provides your custom Resource info before passing to the real SpanExporter

**Reply by rkolesnev** [reply]
I reckoned I can just capture 'tenant' attributes as span attributes and write transformer for Collector to do similar overwriting, but SpanExporter approach sounds like a cleaner solution - will dig into it.

Cheers,
Roman
